packages/ui-components/src/components/PortalProvider/PortalProvider.component.js

```diff
@@ -1,79 +1,142 @@
+ *  ```
+ *   <PortalProvider.Portal>
+ *     <MyComponent />
+ *   </PortalProvider.Portal>
+ *  ```
+ */
+const Portal = ({ children = null }) => {
+  const rootRef = useContext(PortalContext)
+  const [isMounted, setIsMounted] = useState(false)
+
+  useEffect(() => {
+    if (rootRef?.current) {
+      setIsMounted(true)
+    }
+  }, [rootRef])
+
+  if (!isMounted) {
+    return null
+  }
+
+  const wrappedChildren = (
+    <div className={`juno-portal`} style={portalStyles}>
+      {children}
+    </div>
+  )
+  return createPortal(wrappedChildren, rootRef.current || document.body)
+}
+
+Portal.propTypes = {
+  /** The children to mount in a portal. Typically, these will be menus, modal dialogs, etc. */
+  children: PropTypes.node,
+}
+
+/** A hook that creates a portal container in the current portal root, and returns this newly created container as a node to use in other components:
+ * ```
+ *   const portalRef = usePortalRef()
+ *
+ *   createPortal(<MyComponent />, portalRef ? portalRef : document.body)
+ * ```
+ *  The ref to the portal container element can also be passed as a parameter to components that expect a reference element for positioning, such as Flatpickr / DateTimePickr.
+ */
 export function usePortalRef() {
-  const ref = useContext(PortalContext)
-  const [_, setInitialized] = useState(ref?.current)
+  const rootRef = useContext(PortalContext)
+  const containerRef = useRef(null)
+  const [isReady, setIsReady] = useState(false)
 
   useEffect(() => {
-    if (!ref) {
+    if (!rootRef || !rootRef.current) {
       console.warn(
-        "usePortalRef should be called inside a PortalProvider! You are probably using a component that renders a portal, e.g. Modal or Select. Be sure that your app is wrapped in an AppShellProvider."
+        "usePortalRef must be called inside a PortalProvider. You are probably using a component that renders a portal, e.g. Modal or Select. Make sure your app is wrapped in an AppShellProvider. Alternatively, a PortalProvider can be included manually."
       )
       return
     }
-    if (ref.current) setInitialized(true)
-  }, [ref])
-  return ref?.current
-}
+    // create a portal container element and append it to the portal root container
+    const containerElement = document.createElement("div")
+    containerElement.style.position = "relative"
+    containerElement.style.zIndex = "1"
+    containerElement.classList.add("juno-portal")
+    rootRef.current.append(containerElement)
+    containerRef.current = containerElement
+    // mark the newly created container ready
+    setIsReady(true)
 
-const Portal = ({ children }) => {
-  const ref = usePortalRef()
-  return ref ? createPortal(children, ref) : null
-}
+    return () => {
+      // Clean up the portal element when unmounting:
+      if (containerRef.current && rootRef?.current) {
+        rootRef.current.removeChild(containerRef.current)
+        containerRef.current = null
+      }
+    }
+  }, [rootRef])
```

Comment:
by ading so many logic inside the useEffect you violate this hook rule 
https://github.com/facebook/react/blob/main/packages/eslint-plugin-react-hooks/README.md
Which called **react-hooks/exhaustive-deps**
For more info please check
https://stackoverflow.com/questions/58866796/understanding-the-react-hooks-exhaustive-deps-lint-rule

That has been not enabled by this project yet, but if we do we need to get reed of the code that you provided.